regolith/tools.py

```diff
@@ -1322,3 +1322,63 @@ def is_fully_appointed(person, begin_date, end_date):
             else:
                 print("appointment gap for {} on {}".format(person.get('_id'), str(day)))
     return status
+
+
+def get_grant_amount(grant, ppl_coll, begin_date=None, end_date=None):
```

Comment:
Thanks @adibaejaz , let's start our work here.

Before I look at the tests, lets discuss what this function is supposed to do.  This is captured in the name, and in the docstring.  From the name I see ``get_grant_amount``.  A function with this name I would probably expect to take one argument (grant_identifier) and have it return a float (the amount of the grant).  That is probably not what you have in mind so we probably want to work on this.

Then the docstring says "Parameters" but there is no description of what the function does.  I can maybe piece it together from the parameters and the returns below, but the overall objective is not clear.

Next, it would be helpful to think a bit about the design of the code.  What functions you will make and are they th eright functions.  Code is more modular and reusable if you make functions that do small, granular (and preferrably abstractish) things and then you make functions to do specific things by chaining them together.  

btw, we should issue a warning or better, throw and exception, if someone has concurrent appointments of a different type.  They can't be a pd and a gra at the same time....

Reply:
Thank you Professor!
1. I have edited the docstring to include a description but it might not be very succinct due to the complications of the function itself (which probably suggests I should work on breaking the function down, as you said)
2. As for the name, do you think something like `get_net_grant_amount` or `get_net_grant_budget` be more informative?
3. Do you think the inconsistent appointment type check might be better suited to the `is_fully_appointed` function?  Or should I add it here?
4. I am trying without much success to think of ways to make this function more abstract and/or granular although I completely see the need, especially since so much of it is dependent on the internal properties of the `grant` object.  I'm not sure how to make this more granular, but on making it more abstract, I could possibly redesign the function to take in any object `thing_gives_money` that has a `budget` field, sum over all its `budget_periods`, and deduct values from it by iterating through a list of objects `thing_takes_money` which are appointed on `thing_gives_money`, and then return the amounts over a period of time? Please let me know what you think!

Reply:
Thanks, this is very helpful.  So I thought a bit about this and I think what we are actually trying to do is to build a "burn", which can be thought of as a plot of amount vs. day of each grant

How about the function name ``burn(grant, appointments, begin_date, end_date)``.  

How about we pass in a collection of appointments rather than a people collection, since we don't care about the people.  This means that we can have a function that collects appointments. so a structure a bit like:
```
def collect_appts(people, grant_id, begin_date=None, end_date=None):
    check that either both begin and end date are specified, or neither.
    apps = []
    for person in people:
       for appt in person['appts']:
           if appt[grant] == grant_id:
                 if begin_date and end_date:
                        if is_current(appt, begin_date, end_date):
                              apps.append(appt)
                 else:
                       apps.append(appt)
   ....
   return appts

def burn(grant_coll, appts_coll, begin_date, end_date):
    burn, day = [], []
    for day in date_range:
         net_load = 0
         for budget in grant:
               add budgeted amounts to net_load
         for appt in appts:
               subtract appt amounts
          check for over-under and do whatever with that info
          burn.append[net_load]
          day.append[day]                     #not sure the best format here...could be just an index but could be a date object?
    return burn, day
          

all_appts = collect_appts(people)
grant1_burn, days = burn(grant1_coll, all_appts, begin_date, end_date)
plot(days, grant1_burn)
```
Of course this will reuse most of your existing code, but just distribute it around a bit, but I think this gives us the flexibility and reusability that we are after.

Reply:
Yes, thank you very much for this! I will work on restructuring the code. However, for the burn function, I think it might be better if we collect all appointments, rather than those only in the date range provided by the user, since the grant amount may depend also on its expenditure before the begin_date provided. So I will still include begin_date and end_date parameters in coll_appts since this would be useful, but will default it to collecting all appointments in the database

Reply:
also a workflow question - is it ok if I work on both of these functions in the same PR? or should I create a new PR for the `collect_appts` function?

Reply:
also Professor, how valuable is it to let the user pass in a `grants_coll` instead of simply a `grant`?  for the purposes of checking overspent/ underspent/ appointment on grant out of money in the `makeappointments` helper, we only need to check one grant at a time. adding a multiple grants functionality complicates the code significantly  with several nested loops and such. if we want to plot the aggregate burn, I could write another function that calls this one for multiple grants, and let the `burn` function tackle one grant at a time?

Reply:
yes, I think you are right on both points.  I was thinking about it being burn per grant, and then total burn is just the sum of these.  Not sure why I put grant_coll there.

Reply:
done!